File: origin/YuGiOhCard/View/DiceView.m
#import "DiceView.h"
#import "StringConsts.h"

@interface DiceView()

@end

@implementation DiceView

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self makeUI];
    }
    return self;
}

-(void) makeUI {
    self.backgroundColor = [UIColor clearColor];
    
    self.imgDice = [[UIImageView alloc] initWithFrame:CGRectMake(50, 0, self.frame.size.width-100, self.frame.size.height)];
    [self.imgDice setContentMode:UIViewContentModeScaleAspectFit];
    [self.imgDice setUserInteractionEnabled:YES];
    UITapGestureRecognizer * singleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(closeClicked:)];
    [self.imgDice addGestureRecognizer:singleTap];
    
    int diceval = arc4random() % 6 + 1;
    NSString * imgName = [NSString stringWithFormat:@"dice%d", diceval];
    [self.imgDice setImage:[UIImage imageNamed:imgName]];
    
    [self addSubview:self.imgDice];
}

-(void) closeClicked: (id) sender {
    if ([self.delegate respondsToSelector:@selector(doneDice)]) {
        [self.delegate doneDice];
    }
}

@end

File: swift/YuGiOhCard/View/DiceView.swift
import UIKit

@objc
protocol DiceDelegate: NSObjectProtocol {
    optional func doneDice()
}

class DiceView: UIView {
    var imgDice: UIImageView?
    var delegate: DiceDelegate?

    override init(frame: CGRect) {
        super.init(frame: frame)
        self.makeUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.makeUI()
    }
    
    func makeUI() {
        self.backgroundColor = UIColor.clearColor()
        self.imgDice = UIImageView(frame: CGRectMake(50, 0, self.frame.size.width-100, self.frame.size.height))
        self.imgDice!.contentMode = UIViewContentMode.ScaleAspectFit
        self.imgDice!.userInteractionEnabled = true
        let singleTap = UITapGestureRecognizer(target: self, action: "closeClicked:")
        self.imgDice!.addGestureRecognizer(singleTap)
    
        let diceval = arc4random() % 6 + 1
        let imgName = "dice\(diceval)"
        self.imgDice!.image = UIImage(named: imgName)
    
        self.addSubview(self.imgDice!)
    }
    
    func closeClicked(sender: AnyObject) {
        self.delegate?.doneDice?()
    }
    

}
